zookeeper-server/src/main/java/org/apache/zookeeper/server/ResponseCache.java

```diff
@@ -0,0 +1,84 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.apache.jute.Record;
+import org.apache.zookeeper.data.Stat;
+
+@SuppressWarnings("serial")
+public class ResponseCache {
+    private static final int DEFAULT_RESPONSE_CACHE_SIZE = 400;
+
+    private static class Entry {
+        public Stat stat;
+        public byte[] data;
+    }
+
+    private Map<String, Entry> cache = Collections.synchronizedMap(
+        new LRUCache<String, Entry>(getResponseCacheSize()));
+
+    public ResponseCache() {
+    }
+
+    public void put(String path, byte[] data, Stat stat) {
+        Entry entry = new Entry();
+        entry.data = data;
+        entry.stat = stat;
+        cache.put(path, entry);
+    }
+
+    public byte[] get(String key, Stat stat) {
+        Entry entry = cache.get(key);
+        if (entry == null) {
+            return null;
+        }
+        if (!stat.equals(entry.stat)) {
+            // The node has been modified, invalidate cache.
+            cache.remove(key);
+            return null;
+        } else {
+            return entry.data;
+        }
+    }
+
+    private static int getResponseCacheSize() {
+        String value = System.getProperty("zookeeper.maxResponseCacheSize");
+        return value == null ? DEFAULT_RESPONSE_CACHE_SIZE : Integer.parseInt(value);
+    }
+
+    public static boolean isEnabled() {
+        return getResponseCacheSize() != 0;
+    }
+
+    private static class LRUCache<K, V> extends LinkedHashMap<K, V> {
+        private int cacheSize;
+
+        public LRUCache(int cacheSize) {
```

Comment:
This constructor will use insertion order, instead of access order. So for cache entries that's most recently queried it's possible these entries get ejected if they were inserted long time ago, when cache capacity is full. I am curious, what's the trade off you made to decide to use insertion order, instead of access order? 

One improvement that could be made is in `ResponseCache.get`, when we have a cache hit, we always remove the entry and re-put the same entry, to keep the most recently queried order up to date.

Reply:
I agreed with @hanm and like the idea of removing and re-putting nodes, but doesn't look like it can be easily implemented.

Reply:
We haven't tested or considered the alternate eviction policies you outline and that LinkedHashMap allows. I see three reasonable paths here, @hanm  @anmolnar  @lvfangmin .

- Merge this pr as it is (perhaps rename LRUCache to just Cache) and open a new JIRA to explore future paths.
- I add another property that lets one toggle between insertion order and access order with the current implementation as the default.
- Drop LinkedHashMap entirely and go with something like a guava Cache.

My preference is to merge something as close to the current implementation as possible so that others can start to enjoy the benefits of the ResponseCache and open a JIRA so the task of extending the cache functionality goes to someone enthusiastic about tackling new feature work. Thoughts?